include/nbl/video/IGPUObjectFromAssetConverter.h

```diff
@@ -591,76 +710,189 @@ auto IGPUObjectFromAssetConverter::create(const asset::ICPUMesh** const _begin,
-                IGPUImageView::SCreationParams tmpViewParams;
-                tmpViewParams.subresourceRange.levelCount = params.mipLevels+1u-lowestPresentMip;
-                // if not all mip levels have been manually specified
-                if (tmpViewParams.subresourceRange.levelCount>1u)
-                {
-                    tmpViewParams.flags = static_cast<IGPUImageView::E_CREATE_FLAGS>(0u);
-                    tmpViewParams.image = core::smart_refctd_ptr(gpuimg);
-                    switch (params.type)
-                    {
-                        case asset::IImage::ET_1D:
-                            tmpViewParams.viewType = IGPUImageView::ET_1D_ARRAY;
-                            break;
-                        case asset::IImage::ET_2D:
-                            if (params.flags & asset::IImage::ECF_CUBE_COMPATIBLE_BIT)
-                              tmpViewParams.viewType = IGPUImageView::ET_CUBE_MAP_ARRAY;
-                            else
-                              tmpViewParams.viewType = IGPUImageView::ET_2D_ARRAY;
-                            break;
-                        case asset::IImage::ET_3D:
-                            tmpViewParams.viewType = IGPUImageView::ET_3D;
-                            break;
-                        default:
-                            assert(false);
-                            break;
-                    }
-                    tmpViewParams.format = params.format;
-                    //tmpViewParams.subresourceRange.aspectMask
-                    tmpViewParams.subresourceRange.baseMipLevel = lowestPresentMip-1u;
-                    tmpViewParams.subresourceRange.layerCount = params.arrayLayers;
-                    auto tmpView = m_driver->createGPUImageView(std::move(tmpViewParams));
-                    // deprecated OpenGL path (do with compute shader in the future)
-                    tmpView->regenerateMipMapLevels();
-                }
+                IGPUCommandBuffer::SImageMemoryBarrier imgbarrier;
+                // TODO layouts
+                imgbarrier.oldLayout = asset::EIL_UNDEFINED;
+                imgbarrier.newLayout = asset::EIL_UNDEFINED;
+                imgbarrier.srcQueueFamilyIndex = _params.transferQueue->getFamilyIndex();
+                imgbarrier.dstQueueFamilyIndex = _params.computeQueue->getFamilyIndex();
+                imgbarrier.image = gpuimg;
+                asset::IImage::SSubresourceRange subres;
+                subres.baseArrayLayer = 0u;
+                subres.layerCount = params.arrayLayers;
+                subres.baseMipLevel = 0u;
+                subres.levelCount = params.mipLevels;
+                imgbarrier.subresourceRange = subres;
+                asset::SMemoryBarrier membarrier;
+                membarrier.srcAccessMask = asset::EAF_TRANSFER_WRITE_BIT;
+                membarrier.dstAccessMask = asset::EAF_SHADER_READ_BIT;
+                imgbarrier.barrier = membarrier;
```

Comment:
ugh we should probably take the final queue family index AND a boolean `enableConcurrentUsage` (mutally exclusive options) as another parameter to the converter

because we should know whether to create new resources with the concurrent usage option, and if not then what queue family should be the owner at the end.

P.S. We should also take a usage bit mask for buffers and images as a parameter (sane default is all bits set) so that the final ownership transfer has all the correct usages (and also because soon the buffers on the vulkan backend will need to be created with usage bits).

Reply:
btw this is all talking about a FINAL ownership transfer... after compute mipmap

Reply:
hm, how does this comment relate to those barriers? am i missing something?

Reply:
if you create resources with the enable concurrent usage flag, then you dont need to do a queue family ownership transfer in the image barrier... so yes it relates to this barrier

if you dont, and the resources are intended to be used with some other queue family than the compute one, then you should also transfer the ownership to that queue family at the very end (extra barrier, I guess, I dont remember all the ways to transfer ownership)

Reply:
> if you create resources with the enable concurrent usage flag, then you dont need to do a queue family ownership transfer in the image barrier

i dont have to do ownership transfer, but dont i have to put a barrier to express memory dependency (assert visibility)?